### ISSUE TITLE
Implement Happy Eyeballs algorithm (IPv4 / IPv6 connectivity)

### ISSUE THREAD

**Issue by mittsh** [robbiehanson/CocoaAsyncSocket#315]
Currently `GCDAsyncSocket` resolves hosts to addresses (DNS) automatically, but doesn't choose the right protocol (IPv4 or IPv6). Indeed, some hostnames have both `A` and `AAAA` records, and interfaces support IPv4 and IPv6, so we have choice to connect with either protocols. Though, sometimes servers listens only one of both protocols even if `A` and `AAAA` records are present. e.g. this is the case of the local PHP server (when running `php -S`) that listens on IPv6 only, even if `localhost` has both IPv4 and IPv6 host records.
This causes a problem as we don't know which protocol to use ahead of time, and need to try to connect with either protocol and then try the other if the first fails. This can be done _manually_ using `IPv4Enabled`, `IPv6Enabled` and `IPv4PreferredOverIPv6`, but I think this could be nicely handled by the library as it isn't a simple problem for applications.

I don't have any deep knowledge of IP protocols nor dual-stack applications, but it seems that the [Happy Eyeballs algorithm](http://en.wikipedia.org/wiki/Happy_Eyeballs) is the answer, see [RFC 6555](http://tools.ietf.org/html/rfc6555).

_(note: I need to solve that problem for my app, so I'll try to come up with a solid implementation and do a PR)_

**Comment by alexargo** [comment]
@mittsh Have you done this yet? We're running into a similar problem when testing on an ipv6-only network that uses nat64/dns64 (like Apple will required in early 2016). DNS lookup returns both ipv4/ipv6 addresses and tries to connect on ipv4, which fails if ipv4 is the preferred method (default).

**Comment by mittsh** [comment]
Hi Alex, I haven't done this so far. Our current implementation is very basic and just tried to connect with IPv4 first and if it fails tried IPv6. I'd be happy to work on this one with you. I'll do some research about this problem.



Micha Mazaheri+1 (415) 326-4242

+33 6 40 52 14 57

**Comment by mittsh** [comment]
Hi @alexargo, I haven't done this so far. Our current implementation is very basic and just tried to connect with IPv4 first and if it fails tried IPv6. I'd be happy to work on this one with you. I'll do some research about this problem.

There's a fairly easy to implement algorithm from [RFC 6555](http://tools.ietf.org/html/rfc6555):

> 6.  Example Algorithm
> 
>    What follows is the algorithm implemented in Google Chrome and
>    Mozilla Firefox.
> 
>    1.  Call getaddinfo(), which returns a list of IP addresses sorted by
>        the host's address preference policy.
> 
>    2.  Initiate a connection attempt with the first address in that list
>        (e.g., IPv6).
> 
>    3.  If that connection does not complete within a short period of
>        time (Firefox and Chrome use 300 ms), initiate a connection
>        attempt with the first address belonging to the other address
>        family (e.g., IPv4).
> 
>    4.  The first connection that is established is used.  The other
>        connection is discarded.
> 
>    If an algorithm were to cache connection success/failure, the caching
>    would occur after step 4 determined which connection was successful.

Note that, the caching step is mandatory (from [4.  Algorithm Requirements](http://tools.ietf.org/html/rfc6555#section-4)):

> After performing the above procedure, the client learns whether connections to the host's IPv6 or IPv4 address were successful.  The client MUST cache information regarding the outcome of each connection attempt, and it uses that information to avoid thrashing the network with subsequent attempts.

**Comment by alexargo** [comment]
My team is trying to decide between doing this (or having some other less robust solution like you mentioned) with CocoaAsyncSocket or rewriting in our simple use of CocoaAsyncSocket using NSStream. Sounds like NSStream already handles this problem: http://lists.apple.com/archives/ipv6-dev/2011/Jul/msg00009.html ... we burned a DTS ticket to see if there was another way to check which was the correct network, and Apple suggested using NSStream(the stuff we would need is in iOS 8+, Mac OS 10.10+). That's also their suggestion from the "Your App and Next Generation Networks" WWDC talk this year (http://www.asciiwwdc.com/2015/sessions/719). 

So basically, we're trying to decide which has a higher likelihood of success:

1) Writing NSStream-based CocoaAsyncSocket replacement for the subset of things we use CocoaAsyncSocket for that is similar to this sample code: https://developer.apple.com/library/mac/samplecode/SC1236/

2) Getting a good enough understand of CocoaAsyncSocket internal workings to write a good Happy Eyeballs Algorithm

3)A crappy hack that will likely work for MOST users and get us past App Review when IPV6 support is required.


I think we'll probably decide sometime early next week what the path forward is for us.

**Comment by mittsh** [comment]
@alexargo not sure if that may interest you, but we built a thin layer over `GCDAsyncSocket` that wraps it and offers a better interface for our use case. This wrapper is currently responsible to try IPv6 if IPv4 fails, as we are giving priority to IPv4 so far. I believe this should be the easiest place where to implement the algorithm. Let me know if you're interested, I could open source this class.

**Comment by alexargo** [comment]
@mittsh That does sound interesting.

**Comment by praveensharma** [comment]
@mittsh is that in swift by any chance?

**Comment by mittsh** [comment]
@praveensharma CocoaAsyncSocket is in ObjC so why would we implement this in Swift?

**Comment by ogkent** [comment]
> @mittsh we built a thin layer over GCDAsyncSocket that wraps it and offers a better interface for our use case. This wrapper is currently responsible to try IPv6 if IPv4 fails, as we are giving priority to IPv4 so far. I believe this should be the easiest place where to implement the algorithm. Let me know if you're interested, I could open source this class.

Would love to see what you did.

**Comment by Dirk-** [comment]
@mittsh Could you provide this wrapper class for GCDAsyncSocket for better IPv6 support? That would be great.

**Comment by mittsh** [comment]
@Dirk- @ogkent @praveensharma @alexargo I finally took some time to release what we have done for Paw – sorry for the massive delay!

It's still missing some documentation, and proper functional tests, but it should be stable enough. You can use it as a wrapper around GCDAsyncSocket, and it will provide IPv4/IPv6 support out of the box (with caching of the preferred host).

https://github.com/luckymarmot/LMGCDAsyncSocketMiddleware

**Comment by chrisballinger** [comment]
@mittsh Interesting! Ideally we should try to implement this upstream, but this would definitely be a help to many during the transition.

**Comment by mittsh** [comment]
@chrisballinger For us it has been also a way to have cleaner read-to-data/read-to-eof/read-to-length logic. Hopefully it will help some other devs.

**Comment by Dirk-** [comment]
@mittsh Great, I will try this asap. Thank you very much!

**Comment by tedgoddard** [comment]
We're using the `kPreferIPv6` and this seems to work well in a NAT64 environment.

**Comment by mittsh** [comment]
@tedgoddard It won't work if you have a DNS config for the host that returns IPv4 & IPv6 addresses, but the IPv6 server is down. In that case, the lib will connect to the IPv6 address since you asked to prefer v6, and will fail as server is down. It will never try IPv4.
